pkg/plugins/runtime/k8s/controllers/serviceless_name_extractor.go

```diff
@@ -0,0 +1,77 @@
+package controllers
+
+import (
+	"context"
+	"strings"
+
+	kube_apps "k8s.io/api/apps/v1"
+	kube_batch "k8s.io/api/batch/v1"
+	kube_core "k8s.io/api/core/v1"
+	kube_client "sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+type NameExtractor struct {
+	ReplicaSetGetter kube_client.Reader
+	JobGetter        kube_client.Reader
+}
+
+// Getting the name of serviceless pods could be done by getting it owner.
+// If we apply this pattern for all the cases it isn't going to be backward
+// compatible. There are only 2 cases where names are consistent with old way
+// Deployments and CronJob.
+func (n *NameExtractor) Name(ctx context.Context, pod *kube_core.Pod) (string, string, error) {
+	owners := pod.GetObjectMeta().GetOwnerReferences()
+	namespace := pod.Namespace
+	for _, owner := range owners {
+		switch owner.Kind {
+		case "ReplicaSet":
+			rs := &kube_apps.ReplicaSet{}
+			rsKey := kube_client.ObjectKey{Namespace: namespace, Name: owner.Name}
+			if err := n.ReplicaSetGetter.Get(ctx, rsKey, rs); err != nil {
+				return "", "", err
+			}
+			if len(rs.OwnerReferences) == 0 {
+				// backwards compatibility
```

Comment:
what would be a behavior if we were to get rid of backward compatibility? Same for other places in this func marked as "backwards compatibility"

Reply:
We would take the owner from the pod and notice that it is `ReplicaSet`, when the `ReplicaSet` has no owner then we would return the name of `ReplicaSet`.
```
Pod name = test-server-123
ReplicaSet name = test-server
result = test-server
```
Currently, it works:
```
Pod name = test-server-123
ReplicaSet name = test-server
result = test
```

Reply:
ok, I see.

What would be the upgrade path? Whether we break it now or in two versions, a user will have to create two policies

Reply:
1. Identify serviceless pods not managed by CronJob/Deployment (here we might create some script/command example)
2. Create Policies for services with the name of the owner e.g.: pod manged by job, service name = job name
3. Upgrade Kuma

We can also introduce flag to disable the new behaviour.

Reply:
We could, but eventually, we will need to drop this flag.
My point here is that I don't think it's worth keeping backward-compatible behavior if a user needs to migrate anyways in X versions. Flag make sense for experimental/dangerous changes or for performing smooth upgrade/rollback. I can't see how a flag is justified here.

Reply:
True, the flag would just allow disabling and avoid solving the issue now.